### PR TITLE
dma: add a DMA driver for the Atmel SAM0 series.

### PR DESCRIPTION
**Pull request by nzmichaelh**
Tested using the memory++ -> memory++ and memory++ -> peripheral modes.

Depending on what happens with #6305 I'll move the fixup changes to the SoC file.

A follow up patches will switch serial TX to use DMA.

### CODE REVIEW

File: drivers/dma/dma_sam0.c [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+/* Handles DMA interrupts and dispatches to the individual channel */
+static void dma_sam0_isr(void *arg)
+{
+	struct device *dev = arg;
+	struct dma_sam0_device *data = dev->driver_data;
+	int channel;
+	u32_t intstatus = DMAC->INTSTATUS.reg;
+
+	for (channel = 0; intstatus != 0; intstatus >>= 1) {
```
Comment (lpereira) [inline drivers/dma/dma_sam0.c]: Given that there's a small possibility of having a lot of interrupts pending when this function is called, it might be better to use find_lsb_set() here:

```C
while (true) {
    int channel = find_lsb_set(intstatus);

    if (channel == 0) {
        break;
    }

    dma_sam0_channel_isr(dev, data, channel);
    intstatus &= ~BIT(channel);
}
```

Comment (nzmichaelh) [reply drivers/dma/dma_sam0.c]: Hi Leandro.  The Cortex-M0+ doesn't have a clz instruction so find_lsb_set becomes a function call.  I'm not sure it'd be faster than this loop, especially as it early terminates.

Comment (lpereira) [reply drivers/dma/dma_sam0.c]: Ah. LGTM then.

File: include/dma.h [hunk 2]
```diff
@@ -34,6 +34,13 @@ enum dma_channel_direction {
 	PERIPHERAL_TO_MEMORY
 };
 
+/** Valid values for @a source_addr_adj and @a dest_addr_adj */
```
Comment (galak) [inline include/dma.h]: This should be its own commit, and its own PR since you are 'updating' the API.

File: tests/drivers/dma/chan_blen_transfer/src/test_dma.c [hunk 3]
```diff
@@ -26,7 +26,9 @@
 #define DMA_DEVICE_NAME CONFIG_DMA_0_NAME
 #define RX_BUFF_SIZE (48)
 
+__aligned(4)
```
Comment (galak) [inline tests/drivers/dma/chan_blen_transfer/src/test_dma.c]: aren't {dest,source}_data_size 1, so why the alignment need?

I'm guessing we need to figure out a broader solution here.

File: include/dma.h (hunk 2 above)
Comment (nzmichaelh) [reply include/dma.h]: Split out.

File: tests/drivers/dma/chan_blen_transfer/src/test_dma.c (hunk 3 above)
Comment (nzmichaelh) [reply tests/drivers/dma/chan_blen_transfer/src/test_dma.c]: Ah, I put this in so I could experiment with different transfer data sizes.  The test passes as-is, but if you change {source,dest}_data_size to, say, 4, then the buffer isn't aligned and the test fails.

I'll split this out into a different PR and we can discuss there.

**Comment by nzmichaelh** [comment]
Dropped the include/dma and tests/ changes, ready for review.

**Comment by galak** [comment]
@nzmichaelh I think you dropped the driver as well :)

File: include/dma.h (hunk 2 above)
Comment (nzmichaelh) [reply include/dma.h]: annnd now actually split out into #6615

**Comment by nzmichaelh** [comment]
@galak heh, yeah.

I've rebased and done a few other fixups.  This depends on #6615 landing first before it'll build though.

**Comment by nzmichaelh** [comment]
Still blocked on #6615.
